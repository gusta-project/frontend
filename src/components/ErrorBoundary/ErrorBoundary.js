import PropTypes from 'prop-types';
import React, { Component } from 'react';

export default class ErrorBoundary extends Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node
    ]).isRequired
  };

  constructor(props) {
    super(props);

    this.state = {
      hasError: false
    };
  }

  static getDerivedStateFromError() {
    return {
      hasError: true
    };
  }

  render() {
    if (this.state.hasError) {
      return <h1>Something went wrong. Please refresh the page.</h1>;
    }

    return this.props.children;
  }
}
