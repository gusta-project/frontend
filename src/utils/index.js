import dayjs from 'dayjs';

import { getApplication, getAuthorization } from 'selectors/application';

/* eslint-disable react/display-name, react/prop-types */

/**
 * Creates an object containing action constants namespaced under the specified reducer.
 *
 * @param {string} reducer The name of the reducer or module
 * @param {string[]} actions A list of action names
 * @return {object} Object with action name keys and full action string values
 */
export function buildActions(reducer, actions) {
  const result = {};

  for (const action of actions) {
    result[action] = `${reducer}/${action}`;
  }

  return result;
}

/**
 * Interpolate strings from the `params` object into the `url`.
 *
 * @param {object} endpoint Object with `url` and optional `params` object
 */
export const buildUrl = (endpoint) => {
  // this is defined by webpack.DefinePlugin
  const baseUrl = API_URL;
  const { url, params } = endpoint;

  if (!params) {
    return `${baseUrl}${url}`;
  }

  return Object.entries(params).reduce((resultUrl, entry) => {
    const [key, value] = entry;

    return resultUrl.replace(`{${key}}`, value);
  }, `${baseUrl}${url}`);
};

/**
 * Gets the initial Redux state from localStorage and the source code defaults.
 */
export const getInitialState = () => {
  // this needs to be require()d because an import results in a
  // circular dependency
  const { initialState } = require('reducers');

  try {
    const [accessToken, expiration] = [
      localStorage.getItem('accessToken'),
      localStorage.getItem('expiration')
    ];

    if (!accessToken) {
      return initialState;
    }

    const expirationDate = dayjs(JSON.parse(expiration));

    if (!expirationDate.isValid() || expirationDate.isBefore(dayjs())) {
      localStorage.removeItem('accessToken');
      localStorage.removeItem('expiration');
      return initialState;
    }

    return {
      ...initialState,
      application: {
        ...getApplication(initialState),
        authorization: {
          ...getAuthorization(initialState),
          accessToken: JSON.parse(accessToken),
          expiration: expirationDate
        }
      }
    };
  } catch {
    return initialState;
  }
};
