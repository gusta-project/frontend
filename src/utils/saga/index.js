import counterUtilities from './counter';
import pageUtilities from './pager';
import toastUtilities from './toasts';

export default {
  ...counterUtilities,
  ...pageUtilities,
  ...toastUtilities
};
