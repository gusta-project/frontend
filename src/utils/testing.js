import React, { createElement } from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

/**
 * Creates a mock component which takes a set of props specified at runtime.
 *
 * @param {string} name The component/element name (in PascalCase)
 * @param {object} props An object containing props to setup
 * @return {object} Mock component with specified name and props
 */
export const mockComponent = (name, props = {}) => () =>
  createElement(name, props, props.children);

/**
 * Creates a mock component which will expose its props for snapshot testing purposes
 * @param {*} name The component/element name (in PascalCase)
 */
export const mockComponentWithProps = (name) => (props) =>
  createElement(name, props, props.children);

/**
 * Wraps a React component in a <MemoryRouter> suitable for testing.
 *
 * @param {Component} WrappedComponent React component to wrap
 */
export const withMemoryRouter = (WrappedComponent, routerProps = {}) => (
  props
) => (
  <MemoryRouter {...routerProps}>
    <WrappedComponent {...props} />
  </MemoryRouter>
);

/**
 * Wraps a React component in a <Provider> suitable for testing.
 *
 * @param {Component} WrappedComponent React component to wrap
 * @param {object} store Mock store
 */
export const withProvider = (WrappedComponent, store) => (props) => (
  <Provider store={store}>
    <WrappedComponent {...props} />
  </Provider>
);
