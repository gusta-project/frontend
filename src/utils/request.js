import axios from 'axios';
import { buildUrl } from 'utils';
import { call, race, delay } from 'redux-saga/effects';

/**
 * Make an XMLHttpRequest to the specified endpoint with the specified request data.
 *
 * @param {object} endpoint An object with `url` and `method` keys
 * @param {object} data An object to use as the request body
 */
class Request {
  *execute({ endpoint, data, options = {} }) {
    try {
      // ensure the endpoint was supplied
      if (!endpoint) {
        throw new Error('No endpoint provided!');
      }

      // ensure properties required by axios are present
      const { url, method } = endpoint;

      if (!url) {
        throw new Error('Endpoint is missing URL!');
      }

      if (!method) {
        throw new Error('Endpoint is missing method!');
      }

      // default timeout of 30 seconds
      const { timeout = 30000 } = options;
      const requestUrl = yield call(buildUrl, endpoint);

      // start a race between the request and a timer, cancel the loser
      const { response } = yield race({
        response: call(axios, {
          url: requestUrl,
          method,
          data
        }),
        timeout: delay(timeout)
      });

      if (!response) {
        throw new Error('Request timed out!');
      }

      const { status } = response;
      const success = status >= 200 && status < 400;

      if (success) {
        return {
          success,
          response
        };
      } else {
        return {
          success,
          error: response
        };
      }
    } catch (error) {
      return {
        success: false,
        error
      };
    }
  }
}

const instance = new Request();

export default instance;
