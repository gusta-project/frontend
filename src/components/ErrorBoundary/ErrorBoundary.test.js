import React from 'react';
import renderer from 'react-test-renderer';

import ErrorBoundary from './ErrorBoundary';
import { mockComponent } from 'utils/testing';

describe('<ErrorBoundary />', () => {
  const TestComponent = mockComponent('Test');

  it('renders the page if there is no error', () => {
    const component = renderer.create(
      <ErrorBoundary>
        <TestComponent />
      </ErrorBoundary>
    );

    expect(component.root.findByType(TestComponent)).toBeDefined();
  });

  it('renders an error page if there is an error', () => {
    const component = renderer.create(
      <ErrorBoundary>
        <TestComponent />
      </ErrorBoundary>
    );
    const instance = component.getInstance();

    // instance.state.hasError = true;
    instance.setState({ hasError: true });

    expect(component.toJSON()).toMatchSnapshot();
  });
});
