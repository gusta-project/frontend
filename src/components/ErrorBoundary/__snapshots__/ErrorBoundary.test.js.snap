// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorBoundary /> renders an error page if there is an error 1`] = `
<h1>
  Something went wrong. Please refresh the page.
</h1>
`;
